### ISSUE TITLE
[Firestore] FIRWriteBatch commitWithCompletion does not run completion block while in offline mode

### ISSUE THREAD

**Issue by lukegeiger** [firebase/firebase-ios-sdk#478]
* Xcode version: Any
  * Firebase SDK version: N/A
  * Firebase Component: Firestore
  * Component version: N/A

Per the Firestore docs, FIRWriteBatch should commit transactions even when in offline mode, however the completion block of the commitWithCompletion only gets run when the commit is synced with the server. Unless I am totally using it wrong, this use case below results in a totally broken user experience for a user because the loading indicator would never go away if a user committed a batched transaction while offline. What is the workaround for this?

```
[self showLoadingIndicator];
[self.batch commitWithCompletion:^(NSError * _Nullable error) {
[weakself hideLoadingIndicator];
            if (error != nil) {
                NSLog(@"Error writing batch %@", error);
                completion(error);
            } else {
                NSLog(@"Batch write succeeded.");
                completion(nil);
            }
        }];
```

**Comment by wilhuff** [comment]
Use completion handlers on a write only for when you really absolutely care about exactly when something commits to the server.

Instead listen to the affected documents, and listen for metadata. You'll get a callback with hasLocalModifications = false when all changes have been synchronized. The good thing about this method is that it survives restarts of the app without needing to know exactly which writes have completed.

In other words: while listening, every time you get a snapshot just do

```
if (snapshot.metadata.hasPendingWrites) {
  [self showLoadingIndicator];
} else {
  [self hideLoadingIndicator];
}
```

**Comment by lukegeiger** [comment]
Hey @wilhuff thanks for the reply. 

Don't you think your solution seems to be a bit unintuitive? Especially since I think my use case is pretty common, and I can't find anywhere on the docs online that do what you just suggested. Is there any downside of modifying the behavior of commitWithCompletion to run its completion block once it hits the cache? Or maybe even just providing another commitWithCompletion that has two blocks, one for the cache one for the server?

**Comment by morganchen12** [comment]
Batched writes succeed or fail as a unit, so you should be able to get away with listening to a single document within the batch.

The solution isn't the most straightforward. I'll add it to the docs.

**Comment by wilhuff** [comment]
So the basic problem is that while offline the user of your app may close it. Your write will remain queued and will complete when the user returns online. Listening for completion to the write doesn't actually tell you when you're done syncing: it only tells you if this specific write has completed while online.

Specifically: after a restart your completion handler is no longer waiting but if you wait your user to know you're syncing you can't tell them it's happening.

Meanwhile local writes will pretty much always succeed and we already guarantee you can read your own writes out of band so you don't have to wait for the completion handler just to see your writes in the next read.

So the suggestion I made was: connect your view to a [listen on the document](https://firebase.google.com/docs/reference/swift/firebasefirestore/api/reference/Classes/DocumentReference#/c:objc(cs)DocumentReference(im)addSnapshotListenerWithOptions:listener:). Opt into snapshot metadata using [includeMetadataChanges](https://firebase.google.com/docs/reference/swift/firebasefirestore/api/reference/Classes/DocumentListenOptions#/c:objc(cs)DocumentListenOptions(py)includeMetadataChanges).

Now for each snapshot the listener returns update the UI. _Completely independent of any writes you perform_, enable the spinner if `snapshot.metadata.hasPendingWrites` is true, remove it otherwise.

Whenever you perform a write we'll deliver a snapshot containing the values in the write and the metadata that indicates the state of the write.

**Comment by lukegeiger** [comment]
@morganchen12 I think you are incorrect, it seems the addSnapshotListenerWithOptions will still call its completion block for every single document in the batch.

@wilhuff  It's possible I am misunderstanding you, but I don't necessarily want to listen on any document when I perform my batch request. I am updating maybe a half dozen or so different models, and It seems unnecessary to listen to every single snapshot listener on them. I instead just want to listen to my batch request and know when all of its individual requests have finished hitting the cache. 

Also back to the basic use case of hiding and showing the spinner,  I essentially would be calling hideSpinner a half dozen or so times (equal to the amount of updates/sets/writes I have in my batch request (that is if I am understanding you correctly) 

For the sake of getting on the same page, here is some sample code and a bit of how I structured my app. For one, everything firebase in my application has some sort of light weight wrapper class around it for the sole purpose of if there is ever a day I need to migrate off of firebase, I essentially should only have to swap out the implementation of one or two files instead of having to fundamentally rewrite the app.

So as stated above I have a class called HMAPIBatch, which is a wrapper around FIRWriteBatch. Here is its header file

```
NS_ASSUME_NONNULL_BEGIN

typedef void(^HMAPIBatchCompletionBlock)(NSError *error);

@interface HMAPIBatch : NSObject

- (void)set:(HMPersistedObject *)persistedObject;

- (void)update:(HMPersistedObject *)persistedObject;

- (void)remove:(HMPersistedObject *)persistedObject;

- (void)executeBatchOnCompletion:(HMAPIBatchCompletionBlock)completion;

@end

NS_ASSUME_NONNULL_END

```

Here is the implementation of the set function which is basically the same implementation as update and remove

```
- (void)set:(HMPersistedObject *)persistedObject
{
    NSString *endpoint = [self _endpointForClass:persistedObject.class];
        FIRDocumentReference *ref = [[self.defaultFirestore collectionWithPath:endpoint] documentWithPath:persistedObject.uuid];

       [self.batch setData:[MTLJSONAdapter JSONDictionaryFromModel:persistedObject error:nil] forDocument:ref];
}
```

And here is the implementation of the executeBatchOnCompletion

```
- (void)executeBatchOnCompletion:(HMAPIBatchCompletionBlock)completion
{
    [self.batch commitWithCompletion:^(NSError * _Nullable error) {
          completion(error);
    }];
}
```

Here is how I envisioned how to use this class

```
HMAPIBatch *batch = [HMAPIBatch alloc]init];

[batch set:someDataBaseModelObject];
[batch update:someOtherDataBaseModelObject];
[batch remove:someOtherOtherDataBaseModelObject];

[self showLoading];
[batch executeBatchOnCompletion:(NSError *error) {
[weakself hideLoading];
}];
```

Now, @morganchen12 told me about the addSnapshotListenerWithOptions call back which would then change my set, remove, and update implementations to something like this.

```
- (void)set:(HMPersistedObject *)persistedObject
{
    NSString *endpoint = [self _endpointForClass:persistedObject.class];
    
    if (endpoint.length && persistedObject.uuid.length) {
        FIRDocumentReference *ref = [[self.defaultFirestore collectionWithPath:endpoint] documentWithPath:persistedObject.uuid];
        
        FIRDocumentListenOptions *options = [FIRDocumentListenOptions options];
        [options includeMetadataChanges:YES];
        
        [ref addSnapshotListenerWithOptions:options
                                   listener:^(FIRDocumentSnapshot *snapshot, NSError *error){
                                       if (snapshot.metadata.hasPendingWrites) {
                                            // do something here possible enter/leave dispatch queues to coordinate    with others.
                                       }
                                   }];
                                     
        [self.batch setData:[MTLJSONAdapter JSONDictionaryFromModel:persistedObject error:nil] forDocument:ref];
    }
}
```  

So my question now is how can I coordinate all of these listeners apart of each individual snapshot so I know when the entire batch just hits the cache? Not just when each individual item hits the cache.

**Comment by wilhuff** [comment]
I appreciate the effort you're putting into this but I think we've derailed ourselves and need to go back to the beginning on this question.

Why do you want to listen to the completion of the batch commit? You seem sensitive to the fact that it's never going to call you back while the client is offline which indicates you somehow expect some lesser committed state to be useful.

The only other state that's useful is when the batch has been applied locally. We define the batch to apply locally immediately. You don't have to wait for it to be done. The very next operation you perform on your database will be immediately available.

Meanwhile the error case you're listening to can't happen until we've spoken to the server. When just applying locally there are no errors to listen to.

So if you only want to know that the batch has been applied locally you don't need a callback at all and there's no spinner to draw.

**Comment by lukegeiger** [comment]
Hey @wilhuff perhaps we did. 

I want to listen to the completion of the batch because I basically wan't to just know when I can hide/show a loading indicator or even just not block UI in any capacity. I also don't want to fork my code based on if I am offline or not. I want to use the same exact code path while my app is offline as when I am online. 

Which is basically just this

```
[self showSpinner];
    [self.batch commitWithCompletion:^(NSError * _Nullable error) {
[weakself hideSpinner];
          completion(error);
    }];
```
I don't neccesarily think anyone responsible for calling commitWithCompletion should care at all if the app is connected to wifi at all, it should just commit the batch and get some sort of verification that it was saved somewhere.

If I am understanding your last reply correctly, if my app has no internet access I would basically have to fork my code every time I want to commit something based on if I have wifi or not

```

if (offline) {
     [self.batch commitWithCompletion:nil];
} else {
[self showSpinner];
    [self.batch commitWithCompletion:^(NSError * _Nullable error) {
[weakself hideSpinner];
          completion(error);
    }];
}

```

Apologizes for any confusion, as you can tell my app heavily relies on the app working perfect in offline mode.

**Comment by wilhuff** [comment]
The point I'm trying to make is that if you don't specifically need to wait for the write to commit at the server you should just always call this:

```
[batch commitWithCompletion:nil];
[self.delegate doSomething];
```

Firestore internally guarantees ordering so there's nothing your delegate can do that would see the state of the database before the commit has been applied locally.

**Comment by lukegeiger** [comment]
So basically, I should never have to hide/show a loading indicator on a commit then regardless of if I am online or offline? Is there anyway we can update the nullability on the completion block then for commitWithCompletion? currently you have to pass something in which is maybe why I assumed it to always be guaranteed to run.

**Comment by wilhuff** [comment]
PR welcome in FIRWriteBatch.h and .m, otherwise we’ll get to it soon.

This likely is a result of wrapping everything in NS_ASSUME_NONNULL_BEGIN
and we just forgot to specifically call the completion block as nullable

**Comment by lukegeiger** [comment]
Ok, thank you @wilhuff for the quick responses! Much appreciated.

**Comment by btrautmann** [comment]
Since this is still open, I just want to ask one follow-up question. Thanks for this discussion by the way, guys, it was super helpful.

@wilhuff Because we don't receive callbacks until the data has been modified at the *server* level, I'm wondering if we can trust that data validation rules for Firestore will be enforced at the local level as well. That is, if I was to write a corrupt document to some collection, can I always expect an error to be returned immediately, even while offline? Or will that error only be found when sync happens?

**Comment by btrautmann** [comment]
Also, one other potential quirk with this implementation:

When we delete a collection following [this](https://firebase.google.com/docs/firestore/manage-data/delete-data) strategy, we do deletion in 'batches'. However, using `Tasks.await()`, we wait for the batch to finish each time a batch is deleted. Because this callback never happens, only the first batch is ever deleted and so if I have 100 things to delete and I do it in batches of 10, I will have 90 items to delete that won't ever get deleted.

Essentially what I'm asking is, how does one chain together calls when offline? Because no callbacks are triggered, things done in batches or succession are impossible.

**Comment by wilhuff** [comment]
@btrautmann sorry this got lost in the end-of-year shuffle.

Re: rules execution: rules are only ever run on the server. If you write a document that's invalid according to your rules and you intend to take some kind of dependent action on success the only way to do that is to wait for the completion handler to be called. If you're interested in attempting to do this but want to fail if you're offline (because that couldn't ever succeed) then you should use a transaction.

Re: batch deletion offline: I'm not sure I understand the validity of the use case. Deleting a collection while offline requires that every document in the collection be available locally. If that's somehow true in your case (and you're confident that no other client is out there creating documents in the collection) then my advice still stands: you don't need to wait for completion.

In this case iterate over the documents in your collection and create batches of deletes. Once you come back online (which might happen after an app restart) the client will drain its local queue of batches.

**Comment by btrautmann** [comment]
@wilhuff Thanks so much for that answer; no worries on the delay. This makes sense.
